File: swift-gpx/src/main/swift-gen/Gpx.swift

enum FixTypeEnum: String, Codable
{
    case _2d = "2d"
    case _3d = "3d"
    case Dgps = "dgps"
    case None = "none"
    case Pps = "pps"
}

struct Gpx: Codable
{
    let version: String = "1.1"
    var creator: String = ""
    var metadata: Metadata?
    var wpts: [Wpt] = []
    var rtes: [Rte] = []
    var trks: [Trk] = []
    var extensions: [String:String] = [:]

}

struct Metadata: Codable
{
    var name: String?
    var desc: String?
    var author: Person?
    var copyright: Copyright?
    var links: [Link] = []
    var time: Date?
    var keywords: String?
    var bounds: Bounds?
    var extensions: [String:String] = [:]

}

struct Wpt: Codable
{
    var lat: Double = 0.0
    var lon: Double = 0.0
    var ele: Double?
    var time: Date?
    var magvar: Double?
    var geoidheight: Double?
    var name: String?
    var cmt: String?
    var desc: String?
    var src: String?
    var links: [Link] = []
    var sym: String?
    var type: String?
    var fix: FixTypeEnum?
    var sat: Int32?
    var hdop: Double?
    var vdop: Double?
    var pdop: Double?
    var ageofdgpsdata: Double?
    var dgpsid: Int?
    var extensions: [String:String] = [:]

}

struct Rte: Codable
{
    var name: String?
    var cmt: String?
    var desc: String?
    var src: String?
    var links: [Link] = []
    var number: Int32?
    var type: String?
    var extensions: [String:String] = [:]
    var rtepts: [Wpt] = []

}

struct Trk: Codable
{
    var name: String?
    var cmt: String?
    var desc: String?
    var src: String?
    var links: [Link] = []
    var number: Int32?
    var type: String?
    var extensions: [String:String] = [:]
    var trksegs: [Trkseg] = []

}

struct Trkseg: Codable
{
    var trkpts: [Wpt] = []
    var extensions: [String] = []

}

struct Copyright: Codable
{
    var author: String = ""
    var year: Int?
    var license: String?

}

struct Link: Codable
{
    var href: String = ""
    var text: String?
    var type: String?

}

struct Email: Codable
{
    var id: String = ""
    var domain: String = ""

}

struct Person: Codable
{
    var name: String?
    var email: Email?
    var link: Link?

}

struct Pt: Codable
{
    var lat: Double = 0.0
    var lon: Double = 0.0
    var ele: Double?
    var time: Date?

}

struct Bounds: Codable
{
    var minlat: Double = 0.0
    var minlon: Double = 0.0
    var maxlat: Double = 0.0
    var maxlon: Double = 0.0

}

extension Gpx: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + version.hashValue
        result = 31 * result + creator.hashValue
        if let metadata = metadata { result = 31 * result + metadata.hashValue }
        result = 31 * result + wpts.count
        result = 31 * result + rtes.count
        result = 31 * result + trks.count
        result = 31 * result + extensions.count
        return result;
    }

    static func ==(rhs: Gpx, lhs: Gpx) -> Bool {
        guard lhs.version == rhs.version else { return false }
        guard lhs.creator == rhs.creator else { return false }
        guard lhs.metadata == rhs.metadata else { return false }
        guard lhs.wpts == rhs.wpts else { return false }
        guard lhs.rtes == rhs.rtes else { return false }
        guard lhs.trks == rhs.trks else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        return true;
    }
}

extension Metadata: Hashable
{
    var hashValue: Int {
        var result = 1
        if let name = name { result = 31 * result + name.hashValue }
        if let desc = desc { result = 31 * result + desc.hashValue }
        if let author = author { result = 31 * result + author.hashValue }
        if let copyright = copyright { result = 31 * result + copyright.hashValue }
        result = 31 * result + links.count
        if let time = time { result = 31 * result + time.hashValue }
        if let keywords = keywords { result = 31 * result + keywords.hashValue }
        if let bounds = bounds { result = 31 * result + bounds.hashValue }
        result = 31 * result + extensions.count
        return result;
    }

    static func ==(rhs: Metadata, lhs: Metadata) -> Bool {
        guard lhs.name == rhs.name else { return false }
        guard lhs.desc == rhs.desc else { return false }
        guard lhs.author == rhs.author else { return false }
        guard lhs.copyright == rhs.copyright else { return false }
        guard lhs.links == rhs.links else { return false }
        guard lhs.time == rhs.time else { return false }
        guard lhs.keywords == rhs.keywords else { return false }
        guard lhs.bounds == rhs.bounds else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        return true;
    }
}

extension Wpt: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + lat.hashValue
        result = 31 * result + lon.hashValue
        if let ele = ele { result = 31 * result + ele.hashValue }
        if let time = time { result = 31 * result + time.hashValue }
        if let magvar = magvar { result = 31 * result + magvar.hashValue }
        if let geoidheight = geoidheight { result = 31 * result + geoidheight.hashValue }
        if let name = name { result = 31 * result + name.hashValue }
        if let cmt = cmt { result = 31 * result + cmt.hashValue }
        if let desc = desc { result = 31 * result + desc.hashValue }
        if let src = src { result = 31 * result + src.hashValue }
        result = 31 * result + links.count
        if let sym = sym { result = 31 * result + sym.hashValue }
        if let type = type { result = 31 * result + type.hashValue }
        if let fix = fix { result = 31 * result + fix.hashValue }
        if let sat = sat { result = 31 * result + sat.hashValue }
        if let hdop = hdop { result = 31 * result + hdop.hashValue }
        if let vdop = vdop { result = 31 * result + vdop.hashValue }
        if let pdop = pdop { result = 31 * result + pdop.hashValue }
        if let ageofdgpsdata = ageofdgpsdata { result = 31 * result + ageofdgpsdata.hashValue }
        if let dgpsid = dgpsid { result = 31 * result + dgpsid.hashValue }
        result = 31 * result + extensions.count
        return result;
    }

    static func ==(rhs: Wpt, lhs: Wpt) -> Bool {
        guard lhs.lat == rhs.lat else { return false }
        guard lhs.lon == rhs.lon else { return false }
        guard lhs.ele == rhs.ele else { return false }
        guard lhs.time == rhs.time else { return false }
        guard lhs.magvar == rhs.magvar else { return false }
        guard lhs.geoidheight == rhs.geoidheight else { return false }
        guard lhs.name == rhs.name else { return false }
        guard lhs.cmt == rhs.cmt else { return false }
        guard lhs.desc == rhs.desc else { return false }
        guard lhs.src == rhs.src else { return false }
        guard lhs.links == rhs.links else { return false }
        guard lhs.sym == rhs.sym else { return false }
        guard lhs.type == rhs.type else { return false }
        guard lhs.fix == rhs.fix else { return false }
        guard lhs.sat == rhs.sat else { return false }
        guard lhs.hdop == rhs.hdop else { return false }
        guard lhs.vdop == rhs.vdop else { return false }
        guard lhs.pdop == rhs.pdop else { return false }
        guard lhs.ageofdgpsdata == rhs.ageofdgpsdata else { return false }
        guard lhs.dgpsid == rhs.dgpsid else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        return true;
    }
}

extension Rte: Hashable
{
    var hashValue: Int {
        var result = 1
        if let name = name { result = 31 * result + name.hashValue }
        if let cmt = cmt { result = 31 * result + cmt.hashValue }
        if let desc = desc { result = 31 * result + desc.hashValue }
        if let src = src { result = 31 * result + src.hashValue }
        result = 31 * result + links.count
        if let number = number { result = 31 * result + number.hashValue }
        if let type = type { result = 31 * result + type.hashValue }
        result = 31 * result + extensions.count
        result = 31 * result + rtepts.count
        return result;
    }

    static func ==(rhs: Rte, lhs: Rte) -> Bool {
        guard lhs.name == rhs.name else { return false }
        guard lhs.cmt == rhs.cmt else { return false }
        guard lhs.desc == rhs.desc else { return false }
        guard lhs.src == rhs.src else { return false }
        guard lhs.links == rhs.links else { return false }
        guard lhs.number == rhs.number else { return false }
        guard lhs.type == rhs.type else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        guard lhs.rtepts == rhs.rtepts else { return false }
        return true;
    }
}

extension Trk: Hashable
{
    var hashValue: Int {
        var result = 1
        if let name = name { result = 31 * result + name.hashValue }
        if let cmt = cmt { result = 31 * result + cmt.hashValue }
        if let desc = desc { result = 31 * result + desc.hashValue }
        if let src = src { result = 31 * result + src.hashValue }
        result = 31 * result + links.count
        if let number = number { result = 31 * result + number.hashValue }
        if let type = type { result = 31 * result + type.hashValue }
        result = 31 * result + extensions.count
        result = 31 * result + trksegs.count
        return result;
    }

    static func ==(rhs: Trk, lhs: Trk) -> Bool {
        guard lhs.name == rhs.name else { return false }
        guard lhs.cmt == rhs.cmt else { return false }
        guard lhs.desc == rhs.desc else { return false }
        guard lhs.src == rhs.src else { return false }
        guard lhs.links == rhs.links else { return false }
        guard lhs.number == rhs.number else { return false }
        guard lhs.type == rhs.type else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        guard lhs.trksegs == rhs.trksegs else { return false }
        return true;
    }
}

extension Trkseg: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + trkpts.count
        result = 31 * result + extensions.count
        return result;
    }

    static func ==(rhs: Trkseg, lhs: Trkseg) -> Bool {
        guard lhs.trkpts == rhs.trkpts else { return false }
        guard lhs.extensions == rhs.extensions else { return false }
        return true;
    }
}

extension Copyright: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + author.hashValue
        if let year = year { result = 31 * result + year.hashValue }
        if let license = license { result = 31 * result + license.hashValue }
        return result;
    }

    static func ==(rhs: Copyright, lhs: Copyright) -> Bool {
        guard lhs.author == rhs.author else { return false }
        guard lhs.year == rhs.year else { return false }
        guard lhs.license == rhs.license else { return false }
        return true;
    }
}

extension Link: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + href.hashValue
        if let text = text { result = 31 * result + text.hashValue }
        if let type = type { result = 31 * result + type.hashValue }
        return result;
    }

    static func ==(rhs: Link, lhs: Link) -> Bool {
        guard lhs.href == rhs.href else { return false }
        guard lhs.text == rhs.text else { return false }
        guard lhs.type == rhs.type else { return false }
        return true;
    }
}

extension Email: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + id.hashValue
        result = 31 * result + domain.hashValue
        return result;
    }

    static func ==(rhs: Email, lhs: Email) -> Bool {
        guard lhs.id == rhs.id else { return false }
        guard lhs.domain == rhs.domain else { return false }
        return true;
    }
}

extension Person: Hashable
{
    var hashValue: Int {
        var result = 1
        if let name = name { result = 31 * result + name.hashValue }
        if let email = email { result = 31 * result + email.hashValue }
        if let link = link { result = 31 * result + link.hashValue }
        return result;
    }

    static func ==(rhs: Person, lhs: Person) -> Bool {
        guard lhs.name == rhs.name else { return false }
        guard lhs.email == rhs.email else { return false }
        guard lhs.link == rhs.link else { return false }
        return true;
    }
}

extension Pt: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + lat.hashValue
        result = 31 * result + lon.hashValue
        if let ele = ele { result = 31 * result + ele.hashValue }
        if let time = time { result = 31 * result + time.hashValue }
        return result;
    }

    static func ==(rhs: Pt, lhs: Pt) -> Bool {
        guard lhs.lat == rhs.lat else { return false }
        guard lhs.lon == rhs.lon else { return false }
        guard lhs.ele == rhs.ele else { return false }
        guard lhs.time == rhs.time else { return false }
        return true;
    }
}

extension Bounds: Hashable
{
    var hashValue: Int {
        var result = 1
        result = 31 * result + minlat.hashValue
        result = 31 * result + minlon.hashValue
        result = 31 * result + maxlat.hashValue
        result = 31 * result + maxlon.hashValue
        return result;
    }

    static func ==(rhs: Bounds, lhs: Bounds) -> Bool {
        guard lhs.minlat == rhs.minlat else { return false }
        guard lhs.minlon == rhs.minlon else { return false }
        guard lhs.maxlat == rhs.maxlat else { return false }
        guard lhs.maxlon == rhs.maxlon else { return false }
        return true;
    }
}

File: dart_gpx/lib/gpx.dart

import 'package:json_annotation/json_annotation.dart';

part 'gpx.g.dart';

enum FixTypeEnum
{
    @JsonKey(name: "2d")
    twoD,
    @JsonKey(name: "3d")
    threeD,
    dgps,
    none,
    pps
}

@JsonSerializable()
class Gpx {
    final String version = "1.1";
    String creator;
    Metadata metadata;
    List<Wpt> wpts;
    List<Rte> rtes;
    List<Trk> trks;
    Extensions extensions;

    Gpx({
        this.creator, 
        this.metadata, 
        this.wpts, 
        this.rtes, 
        this.trks, 
        this.extensions
    });

    factory Gpx.fromJson(Map<String,dynamic> json) => _$GpxFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Gpx &&
        version == o.version &&
        creator == o.creator &&
        metadata == o.metadata &&
        _equalsList(wpts, o.wpts) &&
        _equalsList(rtes, o.rtes) &&
        _equalsList(trks, o.trks) &&
        extensions == o.extensions;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + version?.hashCode ?? 0;
        result = 31 * result + creator?.hashCode ?? 0;
        result = 31 * result + metadata?.hashCode ?? 0;
        result = 31 * result + wpts?.hashCode ?? 0;
        result = 31 * result + rtes?.hashCode ?? 0;
        result = 31 * result + trks?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$GpxToJson(this);
}

@JsonSerializable()
class Metadata {
    String name;
    String desc;
    Person author;
    Copyright copyright;
    List<Link> links;
    DateTime time;
    String keywords;
    Bounds bounds;
    Extensions extensions;

    Metadata({
        this.name, 
        this.desc, 
        this.author, 
        this.copyright, 
        this.links, 
        this.time, 
        this.keywords, 
        this.bounds, 
        this.extensions
    });

    factory Metadata.fromJson(Map<String,dynamic> json) => _$MetadataFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Metadata &&
        name == o.name &&
        desc == o.desc &&
        author == o.author &&
        copyright == o.copyright &&
        _equalsList(links, o.links) &&
        time == o.time &&
        keywords == o.keywords &&
        bounds == o.bounds &&
        extensions == o.extensions;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + name?.hashCode ?? 0;
        result = 31 * result + desc?.hashCode ?? 0;
        result = 31 * result + author?.hashCode ?? 0;
        result = 31 * result + copyright?.hashCode ?? 0;
        result = 31 * result + links?.hashCode ?? 0;
        result = 31 * result + time?.hashCode ?? 0;
        result = 31 * result + keywords?.hashCode ?? 0;
        result = 31 * result + bounds?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$MetadataToJson(this);
}

@JsonSerializable()
class Wpt {
    double lat;
    double lon;
    double ele;
    DateTime time;
    double magvar;
    double geoidheight;
    String name;
    String cmt;
    String desc;
    String src;
    List<Link> links;
    String sym;
    String type;
    FixTypeEnum fix;
    int sat;
    double hdop;
    double vdop;
    double pdop;
    double ageofdgpsdata;
    int dgpsid;
    Extensions extensions;

    Wpt({
        this.lat, 
        this.lon, 
        this.ele, 
        this.time, 
        this.magvar, 
        this.geoidheight, 
        this.name, 
        this.cmt, 
        this.desc, 
        this.src, 
        this.links, 
        this.sym, 
        this.type, 
        this.fix, 
        this.sat, 
        this.hdop, 
        this.vdop, 
        this.pdop, 
        this.ageofdgpsdata, 
        this.dgpsid, 
        this.extensions
    });

    factory Wpt.fromJson(Map<String,dynamic> json) => _$WptFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Wpt &&
        lat == o.lat &&
        lon == o.lon &&
        ele == o.ele &&
        time == o.time &&
        magvar == o.magvar &&
        geoidheight == o.geoidheight &&
        name == o.name &&
        cmt == o.cmt &&
        desc == o.desc &&
        src == o.src &&
        _equalsList(links, o.links) &&
        sym == o.sym &&
        type == o.type &&
        fix == o.fix &&
        sat == o.sat &&
        hdop == o.hdop &&
        vdop == o.vdop &&
        pdop == o.pdop &&
        ageofdgpsdata == o.ageofdgpsdata &&
        dgpsid == o.dgpsid &&
        extensions == o.extensions;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + lat?.hashCode ?? 0;
        result = 31 * result + lon?.hashCode ?? 0;
        result = 31 * result + ele?.hashCode ?? 0;
        result = 31 * result + time?.hashCode ?? 0;
        result = 31 * result + magvar?.hashCode ?? 0;
        result = 31 * result + geoidheight?.hashCode ?? 0;
        result = 31 * result + name?.hashCode ?? 0;
        result = 31 * result + cmt?.hashCode ?? 0;
        result = 31 * result + desc?.hashCode ?? 0;
        result = 31 * result + src?.hashCode ?? 0;
        result = 31 * result + links?.hashCode ?? 0;
        result = 31 * result + sym?.hashCode ?? 0;
        result = 31 * result + type?.hashCode ?? 0;
        result = 31 * result + fix?.hashCode ?? 0;
        result = 31 * result + sat?.hashCode ?? 0;
        result = 31 * result + hdop?.hashCode ?? 0;
        result = 31 * result + vdop?.hashCode ?? 0;
        result = 31 * result + pdop?.hashCode ?? 0;
        result = 31 * result + ageofdgpsdata?.hashCode ?? 0;
        result = 31 * result + dgpsid?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$WptToJson(this);
}

@JsonSerializable()
class Rte {
    String name;
    String cmt;
    String desc;
    String src;
    List<Link> links;
    int number;
    String type;
    Extensions extensions;
    List<Wpt> rtepts;

    Rte({
        this.name, 
        this.cmt, 
        this.desc, 
        this.src, 
        this.links, 
        this.number, 
        this.type, 
        this.extensions, 
        this.rtepts
    });

    factory Rte.fromJson(Map<String,dynamic> json) => _$RteFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Rte &&
        name == o.name &&
        cmt == o.cmt &&
        desc == o.desc &&
        src == o.src &&
        _equalsList(links, o.links) &&
        number == o.number &&
        type == o.type &&
        extensions == o.extensions &&
        _equalsList(rtepts, o.rtepts);

    @override int get hashCode {
        int result = 17;
        result = 31 * result + name?.hashCode ?? 0;
        result = 31 * result + cmt?.hashCode ?? 0;
        result = 31 * result + desc?.hashCode ?? 0;
        result = 31 * result + src?.hashCode ?? 0;
        result = 31 * result + links?.hashCode ?? 0;
        result = 31 * result + number?.hashCode ?? 0;
        result = 31 * result + type?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        result = 31 * result + rtepts?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$RteToJson(this);
}

@JsonSerializable()
class Trk {
    String name;
    String cmt;
    String desc;
    String src;
    List<Link> links;
    int number;
    String type;
    Extensions extensions;
    List<Trkseg> trksegs;

    Trk({
        this.name, 
        this.cmt, 
        this.desc, 
        this.src, 
        this.links, 
        this.number, 
        this.type, 
        this.extensions, 
        this.trksegs
    });

    factory Trk.fromJson(Map<String,dynamic> json) => _$TrkFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Trk &&
        name == o.name &&
        cmt == o.cmt &&
        desc == o.desc &&
        src == o.src &&
        _equalsList(links, o.links) &&
        number == o.number &&
        type == o.type &&
        extensions == o.extensions &&
        _equalsList(trksegs, o.trksegs);

    @override int get hashCode {
        int result = 17;
        result = 31 * result + name?.hashCode ?? 0;
        result = 31 * result + cmt?.hashCode ?? 0;
        result = 31 * result + desc?.hashCode ?? 0;
        result = 31 * result + src?.hashCode ?? 0;
        result = 31 * result + links?.hashCode ?? 0;
        result = 31 * result + number?.hashCode ?? 0;
        result = 31 * result + type?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        result = 31 * result + trksegs?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$TrkToJson(this);
}

@JsonSerializable()
class Extensions {
    Map<String,dynamic> map;

    Extensions({
        this.map
    });

    factory Extensions.fromJson(Map<String,dynamic> json) => _$ExtensionsFromJson(json);

    bool _equalsMap(Map a, Map b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        return a.keys.every( (key) => b.containsKey(key) && a[key] == b[key] );
    }

    @override bool operator ==(dynamic o) => 
        o is Extensions &&
        _equalsMap(map, o.map);

    @override int get hashCode {
        int result = 17;
        result = 31 * result + map?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$ExtensionsToJson(this);
}

@JsonSerializable()
class Trkseg {
    List<Wpt> trkpts;
    Extensions extensions;

    Trkseg({
        this.trkpts, 
        this.extensions
    });

    factory Trkseg.fromJson(Map<String,dynamic> json) => _$TrksegFromJson(json);

    bool _equalsList(List a, List b) {
        if (identical(a, b) || (a==null && b==null)) return true;
        if (a==null || b==null || a.length != b.length) return false;
        var i = 0;
        return a.every((o) { return a[i++] == o; });
    }

    @override bool operator ==(dynamic o) => 
        o is Trkseg &&
        _equalsList(trkpts, o.trkpts) &&
        extensions == o.extensions;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + trkpts?.hashCode ?? 0;
        result = 31 * result + extensions?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$TrksegToJson(this);
}

@JsonSerializable()
class Copyright {
    String author;
    int year;
    Uri license;

    Copyright({
        this.author, 
        this.year, 
        this.license
    });

    factory Copyright.fromJson(Map<String,dynamic> json) => _$CopyrightFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Copyright &&
        author == o.author &&
        year == o.year &&
        license == o.license;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + author?.hashCode ?? 0;
        result = 31 * result + year?.hashCode ?? 0;
        result = 31 * result + license?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$CopyrightToJson(this);
}

@JsonSerializable()
class Link {
    Uri href;
    String text;
    String type;

    Link({
        this.href, 
        this.text, 
        this.type
    });

    factory Link.fromJson(Map<String,dynamic> json) => _$LinkFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Link &&
        href == o.href &&
        text == o.text &&
        type == o.type;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + href?.hashCode ?? 0;
        result = 31 * result + text?.hashCode ?? 0;
        result = 31 * result + type?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$LinkToJson(this);
}

@JsonSerializable()
class Email {
    String id;
    String domain;

    Email({
        this.id, 
        this.domain
    });

    factory Email.fromJson(Map<String,dynamic> json) => _$EmailFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Email &&
        id == o.id &&
        domain == o.domain;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + id?.hashCode ?? 0;
        result = 31 * result + domain?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$EmailToJson(this);
}

@JsonSerializable()
class Person {
    String name;
    Email email;
    Link link;

    Person({
        this.name, 
        this.email, 
        this.link
    });

    factory Person.fromJson(Map<String,dynamic> json) => _$PersonFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Person &&
        name == o.name &&
        email == o.email &&
        link == o.link;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + name?.hashCode ?? 0;
        result = 31 * result + email?.hashCode ?? 0;
        result = 31 * result + link?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$PersonToJson(this);
}

@JsonSerializable()
class Pt {
    double lat;
    double lon;
    double ele;
    DateTime time;

    Pt({
        this.lat, 
        this.lon, 
        this.ele, 
        this.time
    });

    factory Pt.fromJson(Map<String,dynamic> json) => _$PtFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Pt &&
        lat == o.lat &&
        lon == o.lon &&
        ele == o.ele &&
        time == o.time;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + lat?.hashCode ?? 0;
        result = 31 * result + lon?.hashCode ?? 0;
        result = 31 * result + ele?.hashCode ?? 0;
        result = 31 * result + time?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$PtToJson(this);
}

@JsonSerializable()
class Bounds {
    double minlat;
    double minlon;
    double maxlat;
    double maxlon;

    Bounds({
        this.minlat, 
        this.minlon, 
        this.maxlat, 
        this.maxlon
    });

    factory Bounds.fromJson(Map<String,dynamic> json) => _$BoundsFromJson(json);

    @override bool operator ==(dynamic o) => 
        o is Bounds &&
        minlat == o.minlat &&
        minlon == o.minlon &&
        maxlat == o.maxlat &&
        maxlon == o.maxlon;

    @override int get hashCode {
        int result = 17;
        result = 31 * result + minlat?.hashCode ?? 0;
        result = 31 * result + minlon?.hashCode ?? 0;
        result = 31 * result + maxlat?.hashCode ?? 0;
        result = 31 * result + maxlon?.hashCode ?? 0;
        return result;
    }

    Map<String,dynamic> toJson() => _$BoundsToJson(this);
}
